employee-rostering-gwtui/src/test/java/org/optaweb/employeerostering/gwtui/client/pages/shiftroster/ShiftEditFormTest.java

```diff
@@ -0,0 +1,387 @@
+
+    private List<Spot> spotList;
+
+    private List<Employee> employeeList;
+
+    @Before
+    public void setUp() throws Exception {
+        restCallbackAnswers = new LinkedList();
+        spotList = getSpotList();
+        employeeList = getEmployeeList();
+        doReturn(Optional.of(formPopup)).when(popupFactory).getFormPopup(any());
+        doAnswer(v -> {
+            Consumer c = v.getArgument(0);
+            c.accept(restCallbackAnswers.poll());
+            return null;
+        }).when(callbackFactory).onSuccess(any());
+        testedShiftEditForm = spy(new ShiftEditForm(popupFactory,
+                                                    callbackFactory,
+                                                    root,
+                                                    popupTitle,
+                                                    closeButton,
+                                                    cancelButton,
+                                                    from,
+                                                    to,
+                                                    spotSelect,
+                                                    employeeSelect,
+                                                    pinned,
+                                                    deleteButton,
+                                                    applyButton,
+                                                    tenantStore,
+                                                    eventManager,
+                                                    translationService));
+    }
+
+    private List<Spot> getSpotList() {
+        final String[] SPOT_NAMES = {"Spot1", "Spot2", "Spot3", "Spot4"};
+        List<Spot> out = new ArrayList<>(SPOT_NAMES.length);
+        for (int i = 0; i < SPOT_NAMES.length; i++) {
+            Spot spot = new Spot(0, SPOT_NAMES[i], Collections.emptySet());
+            spot.setId((long) i);
+            out.add(spot);
+        }
+        return out;
+    }
+
+    private List<Employee> getEmployeeList() {
+        final String[] EMPLOYEE_NAMES = {"Employee1", "Employee2", "Employee3", "Employee4"};
+        List<Employee> out = new ArrayList<>(EMPLOYEE_NAMES.length);
+        for (int i = 0; i < EMPLOYEE_NAMES.length; i++) {
+            Employee employee = new Employee(0, EMPLOYEE_NAMES[i]);
+            employee.setId((long) i);
+            out.add(employee);
+        }
+        return out;
+    }
+
+    @Test
+    public void testInit() {
+        restCallbackAnswers.add(spotList);
+        restCallbackAnswers.add(employeeList);
+        ShiftView shiftView = new ShiftView(0, spotList.get(2),
+                                            LocalDateTime.of(2012, 10, 1, 9, 0),
+                                            LocalDateTime.of(2012, 10, 1, 14, 0));
+        doReturn(shiftView).when(mockShiftGridObject).getShiftView();
+        doReturn(spotList.get(2)).when(mockShiftGridObject).getSpot();
```

Comment:
Please prefer `when(...).thenReturn(...)` whenever possible.

See `doReturn()` Javadoc:
> Use `doReturn()` in those rare occasions when you cannot use `when(Object)`.
Beware that `when(Object)` is always recommended for stubbing because it is argument type-safe and more readable (especially when stubbing consecutive calls).

Reply:
Done (and refactored all existing tests to use this pattern when possible)